src/confd/src/infix-if-lag.c

```diff
@@ -0,0 +1,162 @@
+			FILE *fp;
+
+			fp = dagger_fopen_current(net, "exit", lagdiff.old, 55, "exit.ip");
+			if (!fp) {
+				err = -EIO;
+				goto fail;
+			}
+
+			fprintf(fp, "link set %s nomaster\n", ifname);
+			fclose(fp);
+
+			return 0;
+		}
+
+		lagname = lagdiff.new;
+	} else {
+		node = lydx_get_descendant(lyd_child(cif), "lag-port", NULL);
+		lagname = lydx_get_cattr(node, "lag");
+		if (!node || !lagname)
+			goto fail; /* done, nothing to do */
+	}
+
+	fprintf(ip, "link set %s down master %s\n", ifname, lagname);
+
+	/* We depend on lag to exist before we can set master ... */
+	err = dagger_add_dep(net, ifname, lagname);
+	if (err)
+		ERROR("%s: unable to add dep to %s", ifname, lagname);
+fail:
+	return err;
+}
+
+int netdag_gen_lag(sr_session_ctx_t *session, struct dagger *net, struct lyd_node *dif,
+		   struct lyd_node *cif, FILE *ip, int add)
+{
+	const char *lagname = lydx_get_cattr(cif, "name");
+	const char *op = add ? "add" : "set";
+	struct lyd_node *lag, *mon;
+	const char *mode;
+	int err = 0;
+
+	lag = lydx_get_descendant(lyd_child(cif), "lag", NULL);
+	if (!lag)
+		return -EINVAL;
+
+	/* Must take lag down if changing mode */
+	if (!add)
+		fprintf(ip, "link set %s down\n", lagname);
+
+	fprintf(ip, "link %s dev %s", op, lagname);
+
+	if (add) {
+		const char *mac = get_phys_addr(cif, NULL);
+
+		if (!mac) {
+			struct json_t *j;
+
+			j = json_object_get(confd.root, "mac-address");
+			if (j)
+				mac = json_string_value(j);
+		}
+		if (mac)
+			fprintf(ip, " address %s", mac);
+
+		/* on failure, fall back to kernel's random mac */
+	}
+
+	fprintf(ip, " type bond");
+
+	mode = lydx_get_cattr(lag, "mode");
+	if (!strcmp(mode, "lacp")) {
+		struct lyd_node *lacp = lydx_get_child(lag, "lacp");
+
+		if (add)
+			fprintf(ip, " mode 802.3ad");
+
+		mode = lydx_get_cattr(lacp, "mode");
+		if (!strcmp(mode, "active"))
+			mode = "on";
+		else
+			mode = "off";
+
+		fprintf(ip, " lacp_rate %s lacp_active %s",
+			lydx_get_cattr(lacp, "rate"), mode);
+	} else {
+		/* XXX: mode hard-coded for now for mv88e6xxxx operation */
+		if (add)
+			fprintf(ip, " mode balance-xor");
+	}
+
+	/*
+	 * No ARP Monitor support yet, and miimon is required for lacp mode.
+	 * The kernel default interval is 100, which is needlessly often for
+	 * us since we always follow the carrier.
+	 *
+	 * Note: debounce only supported by miimon, not the ARP Monitor.
+	 */
+	fprintf(ip, " miimon 1000 use_carrier 1");
+
+	mon = lydx_get_descendant(lyd_child(cif), "link-monitor", "debounce", NULL);
+	if (mon) {
+		const char *msec;
+
+		msec = lydx_get_cattr(mon, "up");
+		if (msec)
+			fprintf(ip, " updelay %s", msec);
+
+		msec = lydx_get_cattr(mon, "down");
+		if (msec)
+			fprintf(ip, " downdelay %s", msec);
+	}
+
+	fputs("\n", ip);
+
+	if (add) {
+		struct lyd_node *node, *cifs;
+
+		cifs = lydx_get_descendant(lyd_parent(cif), "interfaces", "interface", NULL);
+		LYX_LIST_FOR_EACH(cifs, node, "interface")
+			err += lag_gen_ports(net, NULL, node, ip);
+	}
```

Comment:
Is this not taken care of by `netdag_gen_iface()`'s call to `lag_gen_ports()`?

Reply:
Not when `netdag_must_del()` is triggered, this is the passage when we re-attach unmodified ports when recreating a lag that has had its mode changed.

But you're half right, at initial creation when the lag members are modified as well, this will cause duplicate setups.  That's why the code must be idempotent, including the dagger changes I made.  I could not see any other way of fixing this chicken and egg problem.

I'll add a comment about this in the code here for future readers.